prometheus.lua

```diff
@@ -320,260 +140,374 @@ local function check_metric_and_label_names(metric_name, label_names)
-  local l_names = copy_table(label_names)
-  local l_values = copy_table(label_values)
-
-  -- Last bucket. Note, that the label value is "Inf" rather than "+Inf"
-  -- required by Prometheus. This is necessary for this bucket to be the last
-  -- one when all metrics are lexicographically sorted. "Inf" will get replaced
-  -- by "+Inf" in Prometheus:collect().
-  table.insert(l_names, "le")
-  table.insert(l_values, "Inf")
-  self:inc(name .. "_bucket", l_names, l_values, 1)
-
-  local label_count = #l_names
-  for _, bucket in ipairs(self.buckets[name]) do
-    if value <= bucket then
-      -- last label is now "le"
-      l_values[label_count] = self.bucket_format[name]:format(bucket)
-      self:inc(name .. "_bucket", l_names, l_values, 1)
+function Prometheus:init_worker()
+  if resty_counter_lib then
+    local counter_instance, err = resty_counter_lib.new(self.dict_name, self.sync_interval)
+    if err then
+      error(err, 2)
     end
+    self._counter = counter_instance
+    self.resty_counter_used = true
   end
 end
 
--- Delete all metrics in a gauge or counter. If this gauge or counter have labels, it
---   will delete all the metrics with different label values.
-function Prometheus:reset(name)
-  local keys = self.dict:get_keys(0)
-  for _, key in ipairs(keys) do
-    local value, err = self.dict:get(key)
-    if value then
-      local short_name = short_metric_name(key)
-      if name == short_name then
-        self:set_key(key, nil)
-      end
-    else
-      self:log_error("Error getting '", key, "': ", err)
+local function register(self, name, help, label_names, bucket, typ)
+  if not self.initialized then
+    ngx.log(ngx.ERR, "Prometheus module has not been initialized")
+    return
+  end
+
+  local err = check_metric_and_label_names(name, label_names)
+  if err then
+    self:log_error(err)
+    return
+  end
+
+  local name_maybe_historgram = name:gsub("_bucket$", "")
+                                    :gsub("_count$", "")
+                                    :gsub("_sum$", "")
+  if (self.typ ~= TYPE_HISTOGRAM and (
+      self.registry[name] or self.registry[name_maybe_historgram]
+    )) or
+    (self.typ == TYPE_HISTOGRAM and (
+      self.registry[name .. "_count"] or
+      self.registry[name .. "_sum"] or self.registry[name .. "_bucket"]
+    )) then
+
+    self:log_error("Duplicate metric " .. name)
+    return
+  end
+
+  local metric = {
+    name = name,
+    help = help,
+    typ = typ,
+    label_names = label_names,
+    label_count = label_names and #label_names or 0,
+    -- TODO: lru cache with auto ttl?
+    lookup = {},
+    parent = self,
+    -- store a reference for faster lookup
+    _log_error = function(...) self:log_error(...) end,
+    _log_error_kv = function(...) self:log_error_kv(...) end,
+    _dict = self.dict,
+    -- populate functions
+    -- TODO: how does it compare with metatable lookup cpu/memory-ise?
+    del = del,
```

Comment:
I believe `del` only works for counters and gauges, so should be in the section below.